packages/zoe/src/contractFacet.js

```diff
@@ -287,28 +279,38 @@ export function buildRootObject(_vatPowers) {
       return removePurse(issuerRecord);
     });
 
-  function completeOffers(offerHandlesToDrop) {
+  function completeOffers(offerHandlesToDrop, zoeForZcf) {
     assertOffersAreActive(offerHandlesToDrop);
     offerTable.deleteOffers(offerHandlesToDrop);
     return E(zoeForZcf).completeOffers(offerHandlesToDrop);
   }
 
   /** @type {ContractFacet} */
-  const makeContractFacet = zoeService =>
-    harden({
-      reallocate,
-      addNewIssuer,
-      complete: completeOffers,
-      makeInvitation: (offerHook, inviteDesc, options = harden({})) => {
-        const inviteCallback = harden({ invoke: offerHook });
-        return E(zoeForZcf).makeInvitation(inviteCallback, inviteDesc, options);
+  const makeContractFacet = (
+    zoeService,
+    instanceRecord,
+    inviteIssuer,
+    zoeForZcf,
+  ) => {
+    let publicApiInitialized = false;
+
+    return harden({
+      reallocate: (offerHandles, newAllocations) =>
+        reallocate(offerHandles, newAllocations, zoeForZcf),
+      addNewIssuer: (issuerP, keyword) =>
+        addNewIssuer(issuerP, keyword, instanceRecord, zoeForZcf),
+      complete: offerHandlesToDrop =>
+        completeOffers(offerHandlesToDrop, zoeForZcf),
+      makeInvitation: (offerHandler, inviteDesc, options = harden({})) => {
+        const inviteHandler = harden({ invoke: offerHandler });
+        return E(zoeForZcf).makeInvitation(inviteHandler, inviteDesc, options);
       },
       initPublicAPI: newPublicAPI => {
         assert(
-          publicAPI === undefined,
+          !publicApiInitialized,
           details`the publicAPI has already been initialized`,
         );
-        publicAPI = newPublicAPI;
+        publicApiInitialized = true;
```

Comment:
If a contract needs to get the publicAPI back, how is that done? It looks like it is not locally saved?

Reply:
I think any new API to do this should be postponed to a future PR, as it is a feature not used by any of our contracts thus far.

It's debatable (to me) as to whether any contract really needs it, since the contract is the one that added the publicAPI in the first place.

Reply:
The only place the publicAPI is available in the API is in the instanceRecord. (well, okay, and the offerResult.) The instanceRecord that ZCF makes available is the one that was passed in to startContract.  It's passed as a parameter to makeContractFacet, and a few of the methods in contractFacet use it from there.

The instanceRecord given to startContract has the original promise in it.